keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/Stepsize.java

```diff
@@ -0,0 +1,165 @@
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.VertexId;
+import io.improbable.keanu.vertices.dbl.KeanuRandom;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.LogProbGradientCalculator;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Used by NUTS as an epsilon for the leap frog
+ */
+public class Stepsize {
+
+    private static final double STABILISER = 10;
+    private static final double SHRINKAGE_FACTOR = 0.05;
+    private static final double TEND_TO_ZERO_EXPONENT = 0.75;
+    private static final double STARTING_STEPSIZE = 1;
+
+    private double stepsize;
+    private double averageTreeAcceptanceProb;
+    private double averageAcceptanceProb;
+    private double targetAcceptanceProb;
+    private double logStepSizeFrozen;
+    private double adaptCount;
+    private double shrinkageTarget;
+
+    /**
+     * @param stepsize                  the step size
+     * @param targetAcceptanceProb      the target acceptance probability (lower target equates to a higher step size when tuning)
+     * @param adaptCount                the number of samples to adapt for
+     */
+    Stepsize(double stepsize, double targetAcceptanceProb, int adaptCount) {
+        this.averageAcceptanceProb = 0.;
+        this.targetAcceptanceProb = targetAcceptanceProb;
+        this.stepsize = stepsize;
+        this.logStepSizeFrozen = Math.log(stepsize);
+        this.adaptCount = adaptCount;
+        this.shrinkageTarget = Math.log(10 * stepsize);
+        this.averageTreeAcceptanceProb = 0.;
+    }
+
+    /**
+     * Taken from algorithm 4 in https://arxiv.org/pdf/1111.4246.pdf.
+     *
+     * @param position                      the starting position
+     * @param gradient                      the gradient at the starting position
+     * @param vertices                      the vertices
+     * @param probabilisticVertices         the probabilistic vertices
+     * @param logProbGradientCalculator     the log prob gradient calculator
+     * @param initialLogOfMasterP           the initial master log prob
+     * @param random                        the source of randomness
+     *
+     * @return a starting step size
+     */
+    public static double findStartingStepSize(Map<VertexId, DoubleTensor> position,
+                                       Map<VertexId, DoubleTensor> gradient,
+                                       List<Vertex<DoubleTensor>> vertices,
+                                       List<Vertex> probabilisticVertices,
+                                       LogProbGradientCalculator logProbGradientCalculator,
+                                       double initialLogOfMasterP,
+                                       KeanuRandom random) {
+        double stepsize = STARTING_STEPSIZE;
+        Map<VertexId, DoubleTensor> momentums = new HashMap<>();
+        initializeMomentumForEachVertex(vertices, momentums, random);
```

Comment:
you can nuke this method and do

        Map<VertexId, DoubleTensor> momentums = vertices.stream()
            .collect(Collectors.toMap(
                Vertex::getId,
                v -> random.nextGaussian(v.getShape())
            ));